File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res = []
    function dfs (path, used) {
        if (path.length >= nums.length) res.push([...path])
        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue
            used[i] = true
            path.push(nums[i])
            dfs(path, used)
            path.pop()
            used[i] = false
        }
    }
    dfs([], new Array(nums.length).fill(false))
    return res
}

File: 0046-permutations/0046-permutations.py
class Solution:
    def backtrack(self, res, nums, perm, used):
        if len(perm) == len(nums):
            res.append(perm.copy())
            return
        for i in range(len(nums)):
            if used[i]:
                continue
            perm.append(nums[i])
            used[i] = True
            self.backtrack(res, nums, perm, used)
            perm.pop()
            used[i] = False
        
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        self.backtrack(res, nums, [], [False for i in range(len(nums))])
        return res
        